Update README.md (#13)
WICG/declarative-partial-updates: README.md

## This Proposal
The goal of this proposal is to see whether the platform can help with this kind of experience, by introducing composable fragments (views, because the word fragment is used elsewhere...) and declarative partial document updates as a platform (HTML) primitive.

> It consists of 3 part. The first two stand on their own, and the last one uses the first two:

Reviewer: ```suggestion
It consists of three parts. The first two stand on their own, and the last one uses the first two:
```

* Out-of-order streaming ("splice-stream")
* Route matching
* Declarative same-document navigations

> ### Part 1:Out-of-order streaming ("splice-stream")

Reviewer: ```suggestion
### Part 1: Out-of-order streaming ("splice-stream")
```

1. Potentially, we can make this template point to a *position* in the DOM rather than an ID, e.g. one or two CSS selectors and replacing everything between them.
   If we do that, the template does nothing and is discarded if it doesn't have anywhere to stream into.

This is, in essence, a proposed solution for https://github.com/whatwg/html/issues/2791. 

### Part 2: URL matching with views & routes

To effectively match between parts of the document and the state of the app, there is commonly a relationship between the document's URL and active/visible components of the app.
This is commonly referred to in frameworks as "routes".

The proposal here is to make routes a first-class citizen in HTML, and using that (at first) for two things:
1. Wrapping certain parts of the documents that belong to a certain route with a `<view>` element that describes that relationship.
2. CSS at-rule conditions that get activated by route-matching, similar to media-queries:

```html
<!-- Similar shape to importmap -->
<script type=routemap>
  {
    "movies": { "pattern": { "pathname": "/movies/:movie_id" } },
    "about":  { "pattern": "/about" },
    "home":   { "pattern": ["/home", "/"] }

>   ]

Reviewer 2: ```suggestion
  }
```

</script>

<style>
@route home {
  nav {
    display: none;
  }
}
</style>

<body>
   <view match=home>
   </view>
   <view match=about>
   </view>
   <view match=movies>
   </view>
</body>
```

#### Details of `<view match=...>` and `<script type=routemap>`

1. Every route can be one or more `URLPattern` (more precisely, the parameters to the `URLPattern` constructor)
1. A `view` can be described by using CSS. It is merely an element that has `content-visibility: hidden` when the route doesn't match.
   The choice to make it an element makes it clear semantically that this part of the document is an app fragment.
1. Views can be extended in the future to support a per-view URL ("HTML includes"), and have additional CSS selectors for links that target a particular route.

> 1. Multiple views that match the same route can be present at the same time

Reviewer 2: Similarly, it's possible for multiple routes to match at the same time unless carefully constructed to avoid this. Or should the first or most specific route win so that only one route is considered matching at once?

Author: Yes! Added a note about this. Though we can discuss the specifics.

</section>

<!-- later during the loading process, e.g. when the gallery content is ready -->
<template for="photo-gallery">
  ... the actual gallery...
</template>
```

#### Details of `<template for=...>`

> 1. The `for` attribute behaves like `<label for>`, pointing to an ID.

Reviewer 3: Nit - your enumerations all have "1" rather than "1", "2", "3", ...

This is, in essence, a proposed solution for https://github.com/whatwg/html/issues/2791. 

### Part 2: URL matching with views & routes

To effectively match between parts of the document and the state of the app, there is commonly a relationship between the document's URL and active/visible components of the app.
This is commonly referred to in frameworks as "routes".

The proposal here is to make routes a first-class citizen in HTML, and using that (at first) for two things:
1. Wrapping certain parts of the documents that belong to a certain route with a `<view>` element that describes that relationship.
2. CSS at-rule conditions that get activated by route-matching, similar to media-queries:

```html
<!-- Similar shape to importmap -->
<script type=routemap>
  {
    "movies": { "pattern": { "pathname": "/movies/:movie_id" } },
    "about":  { "pattern": "/about" },
    "home":   { "pattern": ["/home", "/"] }
  ]
</script>

<style>
@route home {
  nav {
    display: none;
  }
}
</style>

<body>
   <view match=home>
   </view>
   <view match=about>
   </view>
   <view match=movies>
   </view>
</body>
```

> #### Details of `<view match=...>` and `<script type=routemap>`

Reviewer: Bikeshedding: since this now uses `routemap` instead of raw URL patterns, perhaps the `match` attribute should be something like `route` (or `matchroute` or `whenroute`)?

That would also be more clear if this were added as a free-standing attribute instead of as part of a new `<view>` element.

1. Every route can be one or more `URLPattern` (more precisely, the paramteres to the `URLPattern` constructor)
1. A `view` can be described by using CSS. It is merely an element that has `content-visibility: hidden` when the route doesn't match.

>    The choice to make it an element makes it clear semantically that this part of the document is an app fragment.

Reviewer: Sorry if this is obvious, but why `content-visibility: hidden` instead of `display: none`?

1. Views can be extended in the future to support a per-view URL ("HTML includes"), and have additional CSS selectors for links that target a particular route.
2. Multiple views that match the same route can be present at the same time

> ## 
> 

Reviewer: ```suggestion
```

</style>
```

In an app that contains matching views, the developer does not have to worry about specifics of the navigation API, or the mechanics of streaming/piping.
All they have to do is list their routes, declare which parts of their document match their URL routes, and then stream the appropriate new content wrapped in `<template for=...>` snippets.

#### Details of declarative interception

> 1. When a route has a `mode: 'same-document'` clause, navigations to matching destinations would be intercepted. The navigation request would still take place, but only `<template for>` elements from the response would be spliced-streamed into the document.

Reviewer 2: Are additional HTTP headers needed for this request, so that the server can optionally respond with just the `<template>` elements for this kind of request?
